orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/ecs/EcsServerGroupCreator.groovy

```diff
@@ -101,4 +93,72 @@ class EcsServerGroupCreator implements ServerGroupCreator, DeploymentDetailsAwar
       return "$repo:$tag"
     }
   }
+
+  private Artifact getTaskDefArtifact(Stage stage, Object input) {
+    TaskDefinitionArtifact taskDefArtifactInput = mapper.convertValue(input, TaskDefinitionArtifact.class)
+
+    Artifact taskDef = artifactResolver.getBoundArtifactForStage(
+      stage,
+      taskDefArtifactInput.artifactId,
+      taskDefArtifactInput.artifact
+    )
+    return taskDef
+  }
+
+  private Map<String, String> getContainerToImageMap(ArrayList<Map<String, Object>> mappings, Stage stage) {
+    def containerToImageMap = [:]
+
+    // each mapping should be in the shape { containerName: "", imageDescription: {}}
+    mappings.each{
+      def imageValue = (Map<String, Object>) it.imageDescription
+      def resolvedImageAddress = getImageAddressFromDescription(imageValue, stage)
+      def name = (String) it.containerName
+      containerToImageMap.put(name, resolvedImageAddress)
+    }
+    return containerToImageMap
+  }
+
+  private String getImageAddressFromDescription(Map<String, Object> description, Stage givenStage) {
+    if (description.fromContext) {
+      if (givenStage.execution.type == ExecutionType.ORCHESTRATION) {
+        // Use image from specific "find image from tags" stage
+        def imageStage = getAncestors(givenStage, givenStage.execution).find {
+          it.refId == description.stageId && it.context.containsKey("amiDetails")
+        }
+
+        if (!imageStage) {
+          throw new IllegalStateException("No image stage found in context for $description.imageLabelOrSha.")
+        }
+
+        description.imageId = imageStage.context.amiDetails.imageId.value.get(0).toString()
+      }
+    }
+
+    if (description.fromTrigger) {
+      if (givenStage.execution.type == ExecutionType.PIPELINE) {
+        def trigger = givenStage.execution.trigger
+
+        if (trigger instanceof DockerTrigger && trigger.account == description.account && trigger.repository == description.repository) {
+          description.tag = trigger.tag
+        }
+
+        description.imageId = buildImageId(description.registry, description.repository, description.tag)
+      }
+
+      if (!description.tag) {
+        throw new IllegalStateException("No tag found for image ${description.registry}/${description.repository} in trigger context.")
+      }
+    }
+
+    if (!description.imageId) {
+      description.imageId = buildImageId(description.registry, description.repository, description.tag)
+    }
+
+    return description.imageId
+  }
+
+  private static class TaskDefinitionArtifact {
+    @Nullable public String artifactId
+    @Nullable public Artifact artifact
```

Comment:
What if these are both null? What is the behavior if the task artifact resolver cant find an artifact?

Reply:
Looks like the resolver will continue try to map it into an object and potentially return a null object. I found that [elsewhere ](https://github.com/spinnaker/orca/blob/master/orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cf/CloudFoundryServerGroupCreator.java#L80) a null check is done after calling it, I can add a similar check for this.